elements/lisk-tree/src/sparse_merkle_tree/sparse_merkle_tree.ts

```diff
@@ -149,10 +149,99 @@ export class SparseMerkleTree {
 		return rootNode;
 	}
 
+	public async remove(key: Buffer): Promise<TreeNode> {
+		if (key.length !== this.keyLength) {
+			throw new Error(`Key is not equal to defined key length of ${this.keyLength}`);
+		}
+
+		const ancestorNodes: TreeNode[] = [];
+		const binaryKey = binaryExpansion(key, this.keyLength);
+		let currentNode = await this.getNode(this._rootHash);
+		let h = 0;
+		let currentNodeSibling: TreeNode = new Empty();
+
+		// Collect all ancestor nodes through traversing the binary expansion by height
+		// End of the loop ancestorNodes has all the branch nodes
+		// currentNode will be the leaf/node we are looking to remove
+		while (currentNode instanceof Branch) {
+			ancestorNodes.push(currentNode);
+			const d = binaryKey[h];
+			if (d === '0') {
+				currentNodeSibling = await this.getNode(currentNode.rightHash);
+				currentNode = await this.getNode(currentNode.leftHash);
+			} else if (d === '1') {
+				currentNodeSibling = await this.getNode(currentNode.leftHash);
+				currentNode = await this.getNode(currentNode.rightHash);
+			}
+			h += 1;
+		}
+
+		// When currentNode is empty, nothing to remove
+		if (currentNode instanceof Empty) {
+			return currentNode;
+		}
+		// When the input key does not match node key, nothing to remove
+		if (!currentNode.key.equals(key)) {
+			return currentNode;
+		}
+		let bottomNode: TreeNode = new Empty();
+
+		// currentNode has a branch sibling, delete currentNode
+		if (currentNodeSibling instanceof Branch) {
+			await this._db.del(currentNode.hash);
+		} else if (currentNodeSibling instanceof Leaf) {
+			// currentNode has a leaf sibling,
+			// remove the leaf and move sibling up the tree
+			await this._db.del(currentNode.hash);
+			bottomNode = currentNodeSibling;
+
+			h -= 1;
+			// In order to move sibling up the tree
+			// an exact emptyHash check is required
+			// not using EMPTY_HASH here to make sure we use correct hash from Empty class
+			const emptyHash = new Empty().hash;
+			while (h > 0) {
+				const p = ancestorNodes[h - 1] as Branch;
+
+				// if one of the children is empty then break the condition
+				if (
+					p instanceof Branch &&
+					!p.leftHash.equals(emptyHash) &&
```

Comment:
another option would be to get the node from left and right hash with `await this.getNode(p.leftHash);` and check using instanceof. This would be more in line with the checks we do before, and avoid having the accessory definition on line 202

Reply:
either way its fine with me, but it will multiple additional call to db? which can be avoided unless the condition is wrong?